articles/151_roslaunch_xml.md

```diff
@@ -0,0 +1,219 @@
+
+##### Examples
+
+```xml
+<include file="/opt/my_launch_file.py"/>
+<include file="$(find-pkg my_pkg)/launch/some_launch_file.xml"/>
+<include file="/opt/my_other_launch_file.xml">
+  <arg name="some_argument" value="dummy_value"/>
+</include>
+```
+
+#### `<group>` Tag
+
+##### Description
+
+The `<group>` tag allows for launch actions' grouping as well as optional launch file configuration scoping.
+
+##### Examples
+
+```xml
+<group scoped="true">
+  <node package="a_ros_package" name="dummy0" ns="my_ns" executable="dummy_node"/>
+  <node package="a_ros_package" name="dummy1" executable="dummy_node"/>
+</group>
+```
+
+#### `<let>` Tag
+
+##### Description
+
+The `<let>` tags allows for definition of scoped launch file configuration variables.
+
+##### Examples
+
+```
+<let var="foo" value="$(env BAR)"/>
+<let var="baz" value="false"/>
+```
+
+#### `<arg>` Tag
+
+##### Description
+
+The `<arg>` tag allows for launch file configuration via the command-line or when including it via an `<include>` tag.
+Arguments are launch configuration variables just like the ones defined by `<let>` tags.
+Arguments are limited to the scope of their definition and thus have to be explictly passed to included files if any.
+
+##### Examples
+
+```xml
+<arg name="publish_frequency" default="10"/>
+<arg name="output_path" description="Output path for some processing pipeline"/>
+```
+
+#### `<node>` Tag
+
+##### Description
+
+The `<node>` tag allows for executing a ROS node in the form of local OS process.
+
+##### Examples
+
+```xml
+<node package="ros_demos" executable="publisher">
+  <param name="publish_frequency" value="10"/>
+  <remap from="generic_topic_name" to="my_topic"/>
+</node>
+```
+
+#### `<param>` Tag
+
+##### Description
+
+The `<param>` tag allows for setting ROS parameters of a ROS node. These
+can be scalar values or sequences of scalar values, defined directly or
+either nested or brought from a YAML file to make a map.
+
+##### Examples
+
+```xml
+<node package="my_pkg" executable="my_node">
+  <param name="some_numeric_param" value="100.2"/>
+  <param name="some_list_param" value="Some phrase,100.0,true" sep=","/>
+  <param name="some_param_group">
+    <param name="some_integer_param" value="10"/>
+  </param>
+  <param from="path/to/param/file.yml"/>
+</node>
+```
+
+#### `<remap>` Tag
+
+##### Description
+
+The `<remap>` tag allows for ROS name remapping of a `<node>` instance.
+
+##### Examples
+
+```xml
+<node package="my_pkg" executable="my_node">
+  <remap from="chatter" to="/my_chatter"/>
+  <remap from="*/stuff" to="private_\1/stuff"/>
+</node>
+```
+
+#### `<env>` Tag
+
+##### Description
+
+The `<env>` tag allows for modifying an OS process environment.
+
+##### Examples
+
+```xml
+<node package="my_pkg" executable="my_node">
+  <env name="RMW_IMPLEMENTATION" value="rmw_fastrtps_cpp"/>
+</node>
+
+<executable cmd="ls" cwd="/var/log">
+  <env name="LD_LIBRARY" value="/lib/some.so"/>
```

Comment:
Can this be used outside of an executable? I think it should be possible.

Reply:
For that, I should add a parsing method to `SetEnviromentVariable` (now it only has meaning inside `node` and `executable`, the schema wasn't up to date).
We have a subtle issue with that, `SetEnviromentVariable` is an action that supports `if` and `unless` condition. But for `node` and `executable`, I was using `additional_env` argument of them, where I cannot use conditions. So exposing `SetEnviromentVariable` variable with the same name would be confusing.

The workarounds I see are the following:
- Document the limitation.
- Change `ExecuteProcess` and `Node` parsing methods. Instead of using `additional_env`, return a sandwich of `SetEnviromentVariable`, `Node/ExecuteProcess`, `UnsetEnvironmentVariable`.
- Use a different tag.

Reply:
You could use `<set_env ...>` outside of execute process and/or node.